### PR TITLE
Handle DnsResolver direct exception on API 34

### PR DESCRIPTION
**Pull request by yschimke**
Failing on master build https://github.com/square/okhttp/actions/runs/7461208679/job/20300867655

Also seen here https://github.com/Ehviewer-Overhauled/Ehviewer/issues/1311

### CODE REVIEW

**Comment by yschimke** [comment]
Can't see an easy way to add a test android.net.DnsResolver is final, and no shadow for robolectric.

**Comment by JakeWharton** [comment]
Robolectric lets you provide your own shadows if you need one it doesn't provide. Should go into `@Config`.

https://robolectric.org/extending/#using-a-custom-shadows

**Comment by yschimke** [comment]
Need to make a few more tests

**Comment by yschimke** [comment]
@JakeWharton care to take another look?

**Comment by JakeWharton** [comment]
I can on Monday

File: okhttp-android/src/main/kotlin/okhttp3/android/AndroidAsyncDns.kt [hunk 1]
```diff
@@ -40,38 +40,47 @@ class AndroidAsyncDns(
...
+            callback.onFailure(
+              hostname,
+              UnknownHostException(e.message).apply {
+                initCause(e)
+              },
+            )
+          }
+        },
+      )
+    } catch (e: Exception) {
```
Comment (swankjesse) [inline okhttp-android/src/main/kotlin/okhttp3/android/AndroidAsyncDns.kt:76]: Any idea what kinds of exceptions we’ll be potentially hiding here? We’re attributing an arbitrary failure like an IllegalStateException as an unknown host

Comment (swankjesse) [reply okhttp-android/src/main/kotlin/okhttp3/android/AndroidAsyncDns.kt:76]: I saw from the attached but report it’s gonna happen when the DNS isn't available.

```
IllegalArgumentException: Network.fromNetworkHandle refusing to instantiate NETID_UNSET Network.
```

I agree with this PR, we’ve got to be resilient to this API that doesn’t offer us a ton of good options.

Comment (yschimke) [reply okhttp-android/src/main/kotlin/okhttp3/android/AndroidAsyncDns.kt:76]: I'll add a comment.

Comment (yschimke) [reply okhttp-android/src/main/kotlin/okhttp3/android/AndroidAsyncDns.kt:76]: https://github.com/square/okhttp/pull/8322/files

**Pull request merged**